src/ray/core_worker/core_worker.cc

```diff
@@ -624,12 +658,13 @@ Status CoreWorker::AllocateReturnObjects(
     bool object_already_exists = false;
     std::shared_ptr<Buffer> data_buffer;
     if (data_sizes[i] > 0) {
-      if (!worker_context_.CurrentActorUseDirectCall()) {
+      if (worker_context_.CurrentActorUseDirectCall() &&
+          data_sizes[i] < kMaxDirectCallObjectSize) {
+        data_buffer = std::make_shared<LocalMemoryBuffer>(data_sizes[i]);
+      } else {
         RAY_RETURN_NOT_OK(
             Create(metadatas[i], data_sizes[i], object_ids[i], &data_buffer));
```

Comment:
Isn't this supposed to be `object_ids[i].WithTransportType(RAYLET)`?

Reply:
I guess it happens to work since all return ids are allocated as RAYLET type already. They are basically ignored by the direct actor transport which then computes its own id with the direct bit set to 1 on the client side when the RPC returns.

Reply:
Ah...can you fix that so we return the right IDs on task submission? Otherwise it's confusing looking at this code why it's correct not to promote the object ID to plasma.

Reply:
Fixing.